### PR TITLE
Manually redo of previous reconnect-buffer PR

### LINKED ISSUE #251: disable sending message during reconnect
hello

how can i disable sending messages after reconnect  that are created during connection lost 

currently published messages are buffered when there is no connection with server and  send immediately after the successful re connection

like this one
https://github.com/nats-io/java-nats/issues/138

### PR DESCRIPTION
**Pull request by danielwertheim**
It was easier to manually reapply changes from PR #284 Also fixes comments in that review. I've opened this PR up as a new PR so that you can compare that I didn't miss anything.

As it replaces #284 it also resolves #251

### CODE REVIEW

File: src/NATS.Client/Conn.cs [hunk 1]
```diff
@@ -2419,6 +2419,28 @@ internal void publish(string subject, string reply, byte[] data, int offset, int
...
+                {
+                    int rbsize = opts.ReconnectBufferSize;
+                    if (rbsize != 0)
+                    {
+                        if (rbsize == -1)
+                        {
+                            throw new NATSReconnectBufferException("Reconnect buffering has been disabled.");
+                        }
+
+                        bw.Flush();
```
Comment (danielwertheim) [inline src/NATS.Client/Conn.cs]: @ColinSullivan1 should this take the flag into consideration?

Comment (ColinSullivan1) [reply src/NATS.Client/Conn.cs]: Yes, it should, good catch.

File: src/NATS.Client/Conn.cs [hunk 2]
```diff
@@ -2419,6 +2419,28 @@ internal void publish(string subject, string reply, byte[] data, int offset, int
...
+
+                        if (pending.Position + count + pubProtoLen > rbsize)
+                        {
+                            throw new NATSReconnectBufferException("Reconnect buffer exceed.");
```
Comment (ColinSullivan1) [inline src/NATS.Client/Conn.cs]: exceed -> exceeded

Comment (danielwertheim) [reply src/NATS.Client/Conn.cs]: Fixed

File: src/NATS.Client/Conn.cs (hunk 1 above)
Comment (danielwertheim) [reply src/NATS.Client/Conn.cs]: Fixed

**Review by ColinSullivan1 — APPROVED** [review]
LGTM - Thanks!

**Comment by danielwertheim** [comment]
@ColinSullivan1 fixed the failing test but I wonder if the "fix" is OK as in, is it OK for it to fail due to increased payload or was the intention to ensure the "queue" is filling up?d

**Comment by ColinSullivan1** [comment]
Hmm...  I wanted to make sure the 2nd publish failed as it would overrun the full buffer, but `PUB foo 20\r\n<20 bytes>\r\n` is 34 bytes, which should have thrown on the first publish?

File: src/NATS.Client/Conn.cs [hunk 3]
```diff
@@ -2419,6 +2419,27 @@ internal void publish(string subject, string reply, byte[] data, int offset, int
...
+                    {
+                        if (rbsize == -1)
+                            throw new NATSReconnectBufferException("Reconnect buffering has been disabled.");
+
+                        if (flushBuffer)
+                            bw.Flush();
+                        else
+                            kickFlusher();
+
+                        if (pending.Position + count + pubProtoLen > rbsize)
```
Comment (danielwertheim) [inline src/NATS.Client/Conn.cs]: @ColinSullivan1  Replying to your Q here instead to get the context.

First publish: `0+20+12 > 32`

Second publish: `0+21+12 > 32`

Comment (ColinSullivan1) [reply src/NATS.Client/Conn.cs]: LGTM.  Thanks!

**Comment by ColinSullivan1** [comment]
LGTM.  Thanks!

**Pull request merged**